server/src-lib/Hasura/RQL/DDL/Remote/Validate.hs

```diff
@@ -141,38 +144,50 @@ validateRemoteArguments expectedArguments providedArguments permittedVariables =
-       when (actualType /= expectedType) (Failure (pure InvalidType)))
-
--- | Produce the type of a value.
-valueType ::
-     HashMap G.Variable FieldInfo
-  -> G.Value
-  -> Validation (NonEmpty ValidationError) (Either G.NamedType G.ListType)
-valueType permittedVariables =
-  \case
+validateType permittedVariables value gType@(getBaseTy -> expectedNamedType) types =
+  case value of
     G.VVariable variable ->
       case HM.lookup variable permittedVariables of
         Nothing -> Failure (pure (InvalidVariable variable permittedVariables))
-        Just fieldInfo -> fmap Left (fieldInfoToNamedType fieldInfo)
-    G.VInt {} -> pure (Left (mkScalarTy PGInteger))
-    G.VFloat {} -> pure (Left (mkScalarTy PGFloat))
-    G.VBoolean {} -> pure (Left (mkScalarTy PGBoolean))
+        Just fieldInfo -> bindValidation
+                            (fieldInfoToNamedType fieldInfo)
+                            (assertType expectedNamedType)
+    G.VInt {} -> assertType (mkScalarTy PGInteger) expectedNamedType
+    G.VFloat {} -> assertType (mkScalarTy PGFloat) expectedNamedType
+    G.VBoolean {} -> assertType (mkScalarTy PGBoolean) expectedNamedType
     G.VNull -> Failure (pure NullNotAllowedHere)
-    G.VString {} -> pure (Left (mkScalarTy PGText))
-    -- TODO: Implement these:
-    -- <https://github.com/tirumaraiselvan/graphql-engine/issues/9>
-    -- Remove the UnsupportedArgumentType constructor when done.
-    value@(G.VEnum _) -> Failure (pure (UnsupportedArgumentType value))
-    value@(G.VList _) -> Failure (pure (UnsupportedArgumentType value))
-    value@(G.VObject _) -> Failure (pure (UnsupportedArgumentType value))
+    G.VString {} -> assertType (mkScalarTy PGText) expectedNamedType
+    v@(G.VEnum _) -> Failure (pure (UnsupportedArgumentType v))
+    G.VList (G.unListValue -> values) -> void $ flip traverse values
+      (\v -> validateType permittedVariables v gType types)
+    G.VObject (G.unObjectValue -> values) -> void $ flip traverse values
+      (\(G.ObjectFieldG name v) ->
+        case HM.lookup expectedNamedType types of
+          Nothing -> Failure (pure $ InputTypeNotFound expectedNamedType )
+          Just typeInfo ->
+            case typeInfo of
+              TIInpObj inpObjTypeInfo ->
+                case HM.lookup name (_iotiFields inpObjTypeInfo) of
+                  Nothing -> Failure (pure $ NoSuchArgumentForRemote name)
+                  Just (_iviType -> expectedType) ->
+                    validateType permittedVariables v expectedType types
+              _                          ->
+                Failure (pure $ InvalidType  "not an input type object"))
+
+assertType ::
+     G.NamedType
+  -> G.NamedType
+  -> Validation (NonEmpty ValidationError) ()
+assertType expectedType actualType =
+  (when (actualType /= expectedType)
+    (Failure (pure (InvalidType $
+                   "expected type: " <> G.showNT expectedType <>
+                   ", received type: " <> G.showNT actualType))))
```

Comment:
I think instead of this we change the `InvalidType` constructor to take two slots e.g.

```haskell
ExpectedTypeButGot !G.NamedType !G.NamedType
```

Then we throw `ExpectedTypeButGot expectedType actualType` so that the caller can make a decision on what to do or how to display the error to the user. Later we might add some context like location or whatever.

Typically this would also be advantageous for a test suite to say `shouldBe mytest (InvalidType .. ..)`, so we're independent of how the error message is pretty printed, but in our case the test suite is in Python.